source/aws_imds_client.c

```diff
@@ -0,0 +1,953 @@
+            }
+        }
+    }
+    aws_mutex_unlock(&client->token_lock);
+
+    /* poll swapped out pending queries if there is any */
+    while (!aws_linked_list_empty(&pending_queries)) {
+        struct aws_linked_list_node *node = aws_linked_list_pop_back(&pending_queries);
+        struct imds_token_query *query = AWS_CONTAINER_OF(node, struct imds_token_query, node);
+        struct imds_user_data *requester = query->user_data;
+        aws_mem_release(client->allocator, query);
+
+        requester->error_code = aws_last_error();
+        if (requester->error_code == AWS_ERROR_SUCCESS) {
+            requester->error_code = AWS_ERROR_UNKNOWN;
+        }
+        s_query_complete(requester);
+    }
+
+    switch (ret) {
+        case AWS_IMDS_TCR_SUCCESS:
+            AWS_LOGF_DEBUG(
+                AWS_LS_IMDS_CLIENT,
+                "(id=%p) IMDS client copied token to requester %p successfully.",
+                (void *)client,
+                (void *)user_data);
+            break;
+
+        case AWS_IMDS_TCR_WAITING_IN_QUEUE:
+            AWS_LOGF_DEBUG(
+                AWS_LS_IMDS_CLIENT, "(id=%p) IMDS client's token is invalid and is now updating.", (void *)client);
+            break;
+
+        case AWS_IMDS_TCR_UNEXPECTED_ERROR:
+            AWS_LOGF_DEBUG(
+                AWS_LS_IMDS_CLIENT,
+                "(id=%p) IMDS client encountered unexpected error when processing token query for requester %p, error: "
+                "%s.",
+                (void *)client,
+                (void *)user_data,
+                aws_error_str(aws_last_error()));
+            break;
+    }
+    return ret;
+}
+
+static void s_invalidate_cached_token_safely(struct imds_user_data *user_data) {
+    bool invalidated = false;
+    struct aws_imds_client *client = user_data->client;
+    aws_mutex_lock(&client->token_lock);
+    if (aws_byte_buf_eq(&user_data->imds_token, &client->cached_token)) {
+        client->token_state = AWS_IMDS_TS_INVALID;
+        invalidated = true;
+    }
+    aws_mutex_unlock(&client->token_lock);
+    if (invalidated) {
+        AWS_LOGF_DEBUG(
+            AWS_LS_IMDS_CLIENT,
+            "(id=%p) IMDS client's cached token is set to be invalid by requester %p.",
+            (void *)client,
+            (void *)user_data);
+    }
+}
+
+/**
+ * Once a requseter returns from token request, it should call this function to unblock all other
+ * waiting requesters. When the token parameter is NULL, means the token request failed. Now we need
+ * a new requester to acquire the token again.
+ */
+static bool s_update_token_safely(struct aws_imds_client *client, struct aws_byte_buf *token, bool token_required) {
+    AWS_FATAL_ASSERT(client);
+    bool updated = false;
+
+    struct aws_linked_list pending_queries;
+    aws_linked_list_init(&pending_queries);
+
+    aws_mutex_lock(&client->token_lock);
+    client->token_required = token_required;
```

Comment:
This feels weird, but I guess it's more of a 3-value state: Unknown, Required, Not required?  I don't know if I'd change anything, it's just odd.

Reply:
It's either V2 required or V1 not required.  He can set either one, client will find it out for him. A more clear setting would reduce unnecessary network request. Maybe we should document the enum better?